File: unity-integration/unity-scripts/StorylineManager.cs
using UnityEngine;
using System;
using System.Collections;
using System.Collections.Generic;

/// <summary>
/// StorylineManager handles interactive stories in the Windgap Academy
/// </summary>
[AddComponentMenu("Windgap/StorylineManager")]
public class StorylineManager : MonoBehaviour
{
    // Reference to other managers
    [SerializeField] private CharacterManager characterManager;
    [SerializeField] private AnimationManager animationManager;
    
    // Story data
    [Serializable]
    public class StoryNode
    {
        public string id;
        public string characterName;
        public string dialogText;
        public string animationName;
        public bool isChoicePoint;
        public List<string> nextNodeIds = new List<string>();
    }
    
    [Serializable]
    public class Story
    {
        public string id;
        public string title;
        public string startNodeId;
        public List<StoryNode> nodes = new List<StoryNode>();
    }
    
    [SerializeField] private List<Story> stories = new List<Story>();
    
    // Current story state
    private Story currentStory;
    private StoryNode currentNode;
    private Coroutine storyCoroutine;
    
    // Dictionary for faster node lookup
    private Dictionary<string, Dictionary<string, StoryNode>> storyNodesMap = new Dictionary<string, Dictionary<string, StoryNode>>();
    
    // Initialize on Awake
    private void Awake()
    {
        // Find references if not set in inspector
        if (characterManager == null)
            characterManager = FindObjectOfType<CharacterManager>();
            
        if (animationManager == null)
            animationManager = FindObjectOfType<AnimationManager>();
            
        // Initialize story nodes dictionary
        InitializeStoryNodesMap();
    }
    
    /// <summary>
    /// Initialize the story nodes dictionary from inspector list
    /// </summary>
    private void InitializeStoryNodesMap()
    {
        foreach (Story story in stories)
        {
            Dictionary<string, StoryNode> nodesDict = new Dictionary<string, StoryNode>();
            
            foreach (StoryNode node in story.nodes)
            {
                nodesDict[node.id] = node;
            }
            
            storyNodesMap[story.id] = nodesDict;
        }
    }
    
    /// <summary>
    /// Start a story by ID
    /// </summary>
    /// <param name="storyId">Story ID to start</param>
    public void StartStory(string storyId)
    {
        // Find the story
        Story story = stories.Find(s => s.id == storyId);
        
        if (story == null)
        {
            Debug.LogWarning($"Story with ID '{storyId}' not found. Using default.");
            story = stories.Count > 0 ? stories[0] : null;
            
            if (story == null)
            {
                Debug.LogError("No stories available.");
                return;
            }
        }
        
        // Stop current story if running
        if (storyCoroutine != null)
        {
            StopCoroutine(storyCoroutine);
            storyCoroutine = null;
        }
        
        // Set current story
        currentStory = story;
        
        // Start story coroutine
        storyCoroutine = StartCoroutine(PlayStoryCoroutine(story.startNodeId));
    }
    
    /// <summary>
    /// Make a choice in the current story
    /// </summary>
    /// <param name="choiceIndex">Index of the choice made</param>
    public void MakeChoice(string choiceIndexStr)
    {
        // Parse choice index
        if (!int.TryParse(choiceIndexStr, out int choiceIndex))
        {
            Debug.LogError($"Invalid choice index: {choiceIndexStr}");
            return;
        }
        
        // Validate current state
        if (currentNode == null || !currentNode.isChoicePoint)
        {
            Debug.LogWarning("No choice point active.");
            return;
        }
        
        // Validate choice index
        if (choiceIndex < 0 || choiceIndex >= currentNode.nextNodeIds.Count)
        {
            Debug.LogError($"Choice index out of range: {choiceIndex}");
            return;
        }
        
        // Get next node ID
        string nextNodeId = currentNode.nextNodeIds[choiceIndex];
        
        // Continue story with the chosen node
        storyCoroutine = StartCoroutine(PlayStoryCoroutine(nextNodeId));
    }
    
    /// <summary>
    /// Play through a story from a given node
    /// </summary>
    private IEnumerator PlayStoryCoroutine(string nodeId)
    {
        // Get the nodes dictionary for the current story
        if (!storyNodesMap.ContainsKey(currentStory.id))
        {
            Debug.LogError($"Story nodes map not found for story: {currentStory.id}");
            yield break;
        }
        
        Dictionary<string, StoryNode> nodesDict = storyNodesMap[currentStory.id];
        
        // Process nodes until the end or a choice point
        bool storyComplete = false;
        
        while (!storyComplete)
        {
            // Get the current node
            if (!nodesDict.ContainsKey(nodeId))
            {
                Debug.LogError($"Node with ID '{nodeId}' not found in story: {currentStory.id}");
                storyComplete = true;
                break;
            }
            
            StoryNode node = nodesDict[nodeId];
            currentNode = node;
            
            // Set active character if specified
            if (!string.IsNullOrEmpty(node.characterName) && characterManager != null)
            {
                characterManager.SetActiveCharacter(node.characterName);
            }
            
            // Play animation if specified
            if (!string.IsNullOrEmpty(node.animationName) && animationManager != null)
            {
                animationManager.PlayAnimation(node.characterName, node.animationName);
            }
            
            // Send node info to React
            if (ReactBridgeManager.Instance != null)
            {
                ReactBridgeManager.Instance.SendToReact("STORY_NODE", node);
            }
            
            // If this is a choice point, wait for player input
            if (node.isChoicePoint)
            {
                break;
            }
            
            // Wait for dialog to be read (approximation based on text length)
            float readTime = Mathf.Max(2.0f, node.dialogText.Length * 0.05f);
            yield return new WaitForSeconds(readTime);
            
            // Move to the next node if available
            if (node.nextNodeIds.Count > 0)
            {
                nodeId = node.nextNodeIds[0];
            }
            else
            {
                storyComplete = true;
            }
        }
        
        // Story complete if we reached the end
        if (storyComplete)
        {
            currentNode = null;
            
            // Notify completion
            if (ReactBridgeManager.Instance != null)
            {
                ReactBridgeManager.Instance.SendToReact("STORY_COMPLETE", new { 
                    storyId = currentStory.id,
                    title = currentStory.title
                });
            }
        }
    }
}

File: unity-integration/StorylineManager.js
// Unity Animation Storylines Loader
// This file is responsible for loading and managing the storyline data for Unity animations

import characterStorylines from "./storylines/character-storylines.json";

/**
 * Class for managing Unity animation storylines
 */
class StorylineManager {
  constructor() {
    this.stories = characterStorylines.stories || [];
    this.currentStory = null;
    this.currentNode = null;
  }

  /**
   * Get a story by ID
   * @param {string} storyId - The ID of the story to retrieve
   * @returns {Object} The story object or null if not found
   */
  getStory(storyId) {
    return this.stories.find((story) => story.id === storyId) || null;
  }

  /**
   * Get all available stories
   * @returns {Array} Array of story objects
   */
  getAllStories() {
    return this.stories;
  }

  /**
   * Get a node from the current story by ID
   * @param {string} nodeId - The ID of the node to retrieve
   * @returns {Object} The node object or null if not found
   */
  getNode(nodeId) {
    if (!this.currentStory) return null;

    return this.currentStory.nodes.find((node) => node.id === nodeId) || null;
  }

  /**
   * Start a story by ID
   * @param {string} storyId - The ID of the story to start
   * @returns {Object} The first node of the story or null if story not found
   */
  startStory(storyId) {
    const story = this.getStory(storyId);
    if (!story) return null;

    this.currentStory = story;
    const startNode = this.getNode(story.startNodeId);
    this.currentNode = startNode;

    return startNode;
  }

  /**
   * Make a choice in the current story
   * @param {number} choiceIndex - The index of the choice made
   * @returns {Object} The next node based on the choice or null if invalid
   */
  makeChoice(choiceIndex) {
    if (!this.currentNode || !this.currentNode.isChoicePoint) return null;
    if (choiceIndex < 0 || choiceIndex >= this.currentNode.nextNodeIds.length) return null;

    const nextNodeId = this.currentNode.nextNodeIds[choiceIndex];
    const nextNode = this.getNode(nextNodeId);
    this.currentNode = nextNode;

    return nextNode;
  }

  /**
   * Advance to the next node in the story
   * @returns {Object} The next node or null if at the end or at a choice point
   */
  advanceStory() {
    if (!this.currentNode) return null;
    if (this.currentNode.isChoicePoint) return null;
    if (this.currentNode.nextNodeIds.length === 0) return null;

    const nextNodeId = this.currentNode.nextNodeIds[0];
    const nextNode = this.getNode(nextNodeId);
    this.currentNode = nextNode;

    return nextNode;
  }

  /**
   * Check if the current story is complete
   * @returns {boolean} True if the story is complete, false otherwise
   */
  isStoryComplete() {
    if (!this.currentNode) return true;
    return this.currentNode.nextNodeIds.length === 0;
  }

  /**
   * Get the title of the current story
   * @returns {string} The title of the current story or null if no story is active
   */
  getCurrentStoryTitle() {
    return this.currentStory ? this.currentStory.title : null;
  }
}

export default new StorylineManager();
